File: src/main/java/com/fengsheng/Recorder.java
package com.fengsheng;

import com.fengsheng.network.ProtoServerChannelHandler;
import com.fengsheng.protos.Errcode;
import com.fengsheng.protos.Fengsheng;
import com.fengsheng.protos.Record;
import com.google.protobuf.ByteString;
import org.apache.log4j.Logger;

import java.io.*;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class Recorder {
    private static final Logger log = Logger.getLogger(Recorder.class);
    private static final short initTocId = ProtoServerChannelHandler.stringHash("init_toc");
    private static final ExecutorService saveLoadPool = Executors.newSingleThreadExecutor();

    private List<Record.recorder_line> list = new ArrayList<>();

    private int currentIndex;

    private volatile boolean loading = false;

    public void add(short messageId, byte[] messageBuf) {
        if (!loading && (messageId == initTocId || list.size() > 0))
            list.add(Record.recorder_line.newBuilder().setNanoTime(System.nanoTime()).setMessageId(messageId)
                    .setMessageBuf(ByteString.copyFrom(messageBuf)).build());
    }

    public void save(Game g, final HumanPlayer p, boolean notify) {
        Date now = new Date();
        LocalDateTime localDateTime = now.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime();
        String timeStr = localDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss"));
        StringBuilder sb = new StringBuilder();
        for (Player player : g.getPlayers()) {
            if (!sb.isEmpty()) sb.append("-");
            sb.append(player.getRoleName());
        }
        final String recordId = Long.toString((now.getTime() / 1000 * 1000 + g.getId() % 100 * 10 + p.location()) % (36L * 36 * 36 * 36 * 36 * 36), Character.MAX_RADIX);
        final String fileName = timeStr + "-" + sb + "-" + p.location() + "-" + recordId;
        var builder = Record.record_file.newBuilder();
        builder.setClientVersion(Config.ClientVersion);
        builder.addAllLines(list);
        Record.record_file recordFile = builder.build();
        saveLoadPool.submit(() -> {
            File file = new File("records/");
            if (!file.exists() && !file.isDirectory() && !file.mkdir())
                log.error("make dir failed: " + file.getName());
            try (DataOutputStream os = new DataOutputStream(new FileOutputStream("records/" + fileName))) {
                os.write(recordFile.toByteArray());
                if (notify) p.send(Fengsheng.save_record_success_toc.newBuilder().setRecordId(recordId).build());
                log.info("save record success: " + recordId);
            } catch (IOException e) {
                log.error("save record failed", e);
            }
        });
    }

    public void load(final int version, final String recordId, final HumanPlayer player) {
        File file = new File("records/");
        if (!file.exists() || !file.isDirectory()) {
            player.send(Errcode.error_code_toc.newBuilder()
                    .setCode(Errcode.error_code.record_not_exists).build());
            return;
        }
        File[] files = recordId.length() == 6 ? file.listFiles((dir, name) -> name.endsWith("-" + recordId)) : null;
        if (files == null || files.length == 0) {
            player.send(Errcode.error_code_toc.newBuilder()
                    .setCode(Errcode.error_code.record_not_exists).build());
            return;
        }
        final File recordFile = files[0];
        loading = true;
        saveLoadPool.submit(() -> {
            try (DataInputStream is = new DataInputStream(new FileInputStream(recordFile))) {
                Record.record_file pb = Record.record_file.parseFrom(is.readAllBytes());
                int recordVersion = pb.getClientVersion();
                if (version < recordVersion) {
                    player.send(Errcode.error_code_toc.newBuilder()
                            .setCode(Errcode.error_code.record_version_not_match)
                            .addIntParams(recordVersion).build());
                    loading = false;
                    return;
                }
                list = pb.getLinesList();
                currentIndex = 0;
                log.info("load record success: " + recordId);
                displayNext(player);
            } catch (IOException e) {
                log.error("load record failed", e);
                player.send(Errcode.error_code_toc.newBuilder()
                        .setCode(Errcode.error_code.load_record_failed).build());
                loading = false;
            }
        });
    }

    private void displayNext(final HumanPlayer player) {
        while (player.isActive()) {
            if (currentIndex >= list.size()) {
                player.send(Fengsheng.display_record_end_toc.getDefaultInstance());
                list = new ArrayList<>();
                loading = false;
                break;
            }
            Record.recorder_line line = list.get(currentIndex);
            player.send((short) line.getMessageId(), line.getMessageBuf().toByteArray());
            if (++currentIndex >= list.size()) {
                player.send(Fengsheng.display_record_end_toc.getDefaultInstance());
                list = new ArrayList<>();
                loading = false;
                break;
            }
            long diffNanoTime = list.get(currentIndex).getNanoTime() - line.getNanoTime();
            if (diffNanoTime > 100000000) {
                GameExecutor.TimeWheel.newTimeout(timeout -> displayNext(player), Math.min(diffNanoTime, 2000000000), TimeUnit.NANOSECONDS);
                break;
            }
        }
    }

    public boolean loading() {
        return loading;
    }
}

File: src/main/kotlin/Recorder.kt
package com.fengsheng

import com.fengsheng.phase.StartGame
import com.fengsheng.phase.WaitForSelectRole
import com.fengsheng.protos.*
import com.fengsheng.protos.Record.record_file
import com.fengsheng.protos.Record.recorder_line
import com.fengsheng.skill.RoleCache
import com.google.protobuf.ByteString
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import org.apache.logging.log4j.kotlin.logger
import java.io.*
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.concurrent.fixedRateTimer

class Recorder {
    private var list: MutableList<recorder_line> = ArrayList()
    private var currentIndex = 0
    private var skipCount = 0

    @Volatile
    var loading = false
        private set

    @Volatile
    private var pausing = false
    fun add(protoName: String, messageBuf: ByteArray?) {
        if (protoName in ignoredProtoNames) return
        if (!loading && ("wait_for_select_role_toc" == protoName || "init_toc" == protoName || list.isNotEmpty())) {
            list.add(recorderLine {
                nanoTime = System.nanoTime()
                this.protoName = protoName
                this.messageBuf = ByteString.copyFrom(messageBuf)
            })
        }
    }

    fun save(g: Game, p: HumanPlayer, notify: Boolean) {
        if (list.size < 10 || g.fsm is StartGame || g.fsm is WaitForSelectRole) return
        val now = Date()
        val localDateTime = now.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime()
        val timeStr = localDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss"))
        val sb = StringBuilder()
        for (player in g.players) {
            if (sb.isNotEmpty()) sb.append("-")
            sb.append(RoleCache.getRoleName(player!!.originRole))
        }
        val recordId = ((now.time / 1000 * 1000 + g.id % 100 * 10 + p.location) %
            (36L * 36 * 36 * 36 * 36 * 36)).toString(Character.MAX_RADIX).addLeadingZero(6)
        val fileName = timeStr + "-" + sb + "-" + p.location + "-" + recordId
        val recordFile = recordFile {
            clientVersion = Config.ClientVersion.get()
            lines.addAll(list)
        }
        saveLoadPool.trySend {
            val file = File("records/")
            if (!file.exists() && !file.isDirectory && !file.mkdir()) logger.error("make dir failed: ${file.name}")
            try {
                DataOutputStream(FileOutputStream("records/$fileName")).use { os ->
                    os.write(recordFile.toByteArray())
                    if (notify) p.send(saveRecordSuccessToc { this.recordId = recordId })
                    logger.info("save record success: $recordId")
                }
            } catch (e: IOException) {
                logger.error("save record failed", e)
            }
        }
    }

    fun load(version: Int, recordId: String, skipCount: Int, player: HumanPlayer) {
        val file = File("records/")
        if (!file.exists() || !file.isDirectory) {
            player.sendErrorMessage("录像不存在")
            return
        }
        val files = if (recordId.length == 6) file.listFiles { _, name -> name.endsWith("-$recordId") } else null
        if (files.isNullOrEmpty()) {
            player.sendErrorMessage("录像不存在")
            return
        }
        val recordFile = files[0]
        this.skipCount = skipCount
        loading = true
        saveLoadPool.trySend {
            try {
                DataInputStream(FileInputStream(recordFile)).use { `is` ->
                    val pb = record_file.parseFrom(`is`.readAllBytes())
                    val recordVersion = pb.clientVersion
                    if (version < recordVersion)
                        logger.warn("客户端版本号过低，录像可能有问题。录像$recordVersion，当前$version")
                    list = pb.linesList
                    currentIndex = 0
                    logger.info("load record success: $recordId")
                    if (player.needWaitLoad)
                        player.send(gameStartToc {})
                    else
                        displayNext(player)
                }
            } catch (e: IOException) {
                logger.error("load record failed", e)
                player.sendErrorMessage("播放录像失败")
                loading = false
            }
        }
    }

    fun pause(pause: Boolean) {
        pausing = pause
    }

    fun displayNext(player: HumanPlayer) {
        if (skipCount > 0) {
            player.send(reconnectToc { isEnd = false })
            while (currentIndex <= skipCount && currentIndex < list.size) {
                val line = list[currentIndex]
                if ("wait_for_select_role_toc" != line.protoName && "select_role_toc" != line.protoName)
                    player.send(line.protoName, line.messageBuf.toByteArray(), true)
                currentIndex++
            }
            player.send(reconnectToc { isEnd = true })
        }
        while (true) {
            if (!player.isActive) {
                loading = false
                return
            }
            if (pausing) {
                GameExecutor.TimeWheel.newTimeout({ displayNext(player) }, 2, TimeUnit.SECONDS)
                return
            }
            if (currentIndex >= list.size) {
                player.send(displayRecordEndToc { })
                list = ArrayList()
                loading = false
                break
            }
            val line = list[currentIndex]
            if ("wait_for_select_role_toc" == line.protoName || "select_role_toc" == line.protoName) {
                currentIndex++
                continue
            }
            player.send(line.protoName, line.messageBuf.toByteArray(), true)
            if (++currentIndex >= list.size) {
                player.send(displayRecordEndToc { })
                list = ArrayList()
                loading = false
                break
            }
            val diffNanoTime = list[currentIndex].nanoTime - line.nanoTime
            if (diffNanoTime > 100000000) {
                var maxInterval = 1000000000L * Config.RecordMaxInterval
                if (list[currentIndex].protoName in fastDisplayProtoNames && line.protoName in fastDisplayProtoNames)
                    maxInterval /= 2
                GameExecutor.TimeWheel.newTimeout(
                    { displayNext(player) },
                    diffNanoTime.coerceAtMost(maxInterval),
                    TimeUnit.NANOSECONDS
                )
                break
            }
        }
    }

    fun reconnect(player: HumanPlayer) {
        player.send(reconnectToc { isEnd = false })
        for (i in 0..(list.size - 2)) {
            val line = list[i]
            player.send(line.protoName, line.messageBuf.toByteArray(), false)
        }
        player.send(reconnectToc { isEnd = true })
        list.lastOrNull()?.let { line ->
            player.send(line.protoName, line.messageBuf.toByteArray(), true)
        }
    }

    companion object {
        private val saveLoadPool = Channel<() -> Unit>(Channel.UNLIMITED)
        private val ignoredProtoNames = listOf(
            "reconnect_toc",
            "heart_toc",
            "game_init_finish_tos",
            "notify_kicked_toc",
            "error_message_toc",
            "notify_player_update_toc"
        )
        private val fastDisplayProtoNames = listOf("notify_phase_toc", "wait_for_cheng_qing_toc")

        init {
            @OptIn(DelicateCoroutinesApi::class)
            GlobalScope.launch {
                while (true) {
                    val f = saveLoadPool.receive()
                    withContext(Dispatchers.IO) { f() }
                }
            }

            fixedRateTimer(daemon = true, period = 600000) {
                val now = Date()
                now.time -= 14 * 24 * 3600 * 1000
                val localDateTime = now.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime()
                val timeStr = localDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss"))
                val file = File("records/")
                if (!file.exists() || !file.isDirectory) {
                    return@fixedRateTimer
                }
                val l = "yyyy-MM-dd-HH-mm-ss".length
                file.listFiles()?.forEach {
                    val fileName = it.name
                    if (fileName.length > l && fileName.substring(0, l) <= timeStr)
                        it.delete()
                }
            }
        }
    }

    private fun String.addLeadingZero(totalLen: Int): String {
        val zeroCount = totalLen - length
        if (zeroCount <= 0) return this
        val sb = StringBuilder()
        repeat(zeroCount) { sb.append('0') }
        sb.append(this)
        return sb.toString()
    }
}
